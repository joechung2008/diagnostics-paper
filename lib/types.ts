export interface BuildInfoProps {
  buildVersion: string;
}

export interface ConfigurationProps {
  config: Record<string, string>;
}

export interface Diagnostics {
  buildInfo: BuildInfoProps;
  extensions: Record<string, Extension>;
  serverInfo: ServerInfoProps;
}

export interface ServerInfoProps {
  deploymentId: string;
  extensionSync: {
    totalSyncAllCount: number;
  };
  hostname: string;
  nodeVersions?: string;
  serverId: string;
  uptime?: number;
}

export interface StageDefinitionProps {
  stageDefinition: Record<string, string[]>;
}

export interface KeyValuePair<TValue> {
  key: string;
  value: TValue;
}

export interface ExtensionInfo {
  extensionName: string;
  manageSdpEnabled: boolean;
  config?: Record<string, string>;
  stageDefinition?: Record<string, string[]>;
}

export interface ExtensionError {
  lastError: {
    errorMessage: string;
    time: string;
  };
}

export type Extension = ExtensionInfo | ExtensionError;

export type ExtensionProps = ExtensionInfo;

export interface ExtensionsProps {
  extensions: Record<string, Extension>;
  onLinkClick(item?: KeyedNavLink): void;
}

export type KeyedNavLink = {
  key: string;
  name: string;
  url?: string;
  [prop: string]: unknown;
};
