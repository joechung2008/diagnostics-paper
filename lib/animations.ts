import { Animated, Easing } from "react-native";

// Animation configurations for fling feedback
export const createFlingResetAnimation = (
  opacity: Animated.Value,
  scale: Animated.Value,
  translateX: Animated.Value
) =>
  Animated.parallel([
    Animated.timing(opacity, {
      duration: 200,
      easing: Easing.out(Easing.quad),
      toValue: 1,
      useNativeDriver: true,
    }),
    Animated.timing(scale, {
      duration: 200,
      easing: Easing.out(Easing.quad),
      toValue: 1,
      useNativeDriver: true,
    }),
    Animated.timing(translateX, {
      duration: 200,
      easing: Easing.out(Easing.quad),
      toValue: 0,
      useNativeDriver: true,
    }),
  ]);

export const createFlingStartAnimation = (
  opacity: Animated.Value,
  scale: Animated.Value
) =>
  Animated.parallel([
    Animated.timing(opacity, {
      toValue: 0.7,
      duration: 150,
      easing: Easing.out(Easing.quad),
      useNativeDriver: true,
    }),
    Animated.timing(scale, {
      toValue: 0.95,
      duration: 150,
      easing: Easing.out(Easing.quad),
      useNativeDriver: true,
    }),
  ]);

export const createFlingDismissAnimation = (
  translateX: Animated.Value,
  targetValue: number
) =>
  Animated.timing(translateX, {
    toValue: targetValue,
    duration: 150,
    easing: Easing.out(Easing.quad),
    useNativeDriver: true,
  });
