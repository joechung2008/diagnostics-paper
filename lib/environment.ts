export type EnvironmentType = (typeof Environment)[keyof typeof Environment];

export const Environment = {
  Public: "https://hosting.portal.azure.net/api/diagnostics",
  Fairfax: "https://hosting.azureportal.usgovcloudapi.net/api/diagnostics",
  Mooncake: "https://hosting.azureportal.chinacloudapi.cn/api/diagnostics",
} as const;

export function getEnvironmentName(
  environment: EnvironmentType | undefined
): string {
  switch (environment) {
    case Environment.Public:
      return "Public Cloud";
    case Environment.Fairfax:
      return "Fairfax";
    case Environment.Mooncake:
      return "Mooncake";
    default:
      return "Select environment";
  }
}
