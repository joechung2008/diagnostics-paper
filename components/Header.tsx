import { useCallback, useMemo, useState } from "react";
import { Appbar, Menu } from "react-native-paper";
import {
  Environment,
  type EnvironmentType,
  getEnvironmentName,
} from "../lib/environment";
import type { Diagnostics, ExtensionInfo } from "../lib/types";
import { isExtensionInfo } from "../lib/utils";

interface HeaderProps {
  diagnostics: Diagnostics | undefined;
  environment: EnvironmentType;
  onEnvironmentChange: (environment: EnvironmentType) => void;
  onExtensionSelect: (extension: ExtensionInfo) => void;
}

const Header = ({
  diagnostics,
  environment,
  onEnvironmentChange,
  onExtensionSelect,
}: HeaderProps) => {
  const [menuVisible, setMenuVisible] = useState(false);

  const environmentName = useMemo(
    () => getEnvironmentName(environment),
    [environment]
  );

  const showPaasServerless = useMemo(
    () => isExtensionInfo(diagnostics?.extensions["paasserverless"]),
    [diagnostics?.extensions]
  );

  const handleEnvironmentChange = useCallback(
    (value: EnvironmentType) => {
      onEnvironmentChange(value);
      setMenuVisible(false);
    },
    [onEnvironmentChange]
  );

  return (
    <Appbar.Header>
      <Menu
        anchor={
          <Appbar.Action
            accessibilityLabel="Select environment"
            icon="menu"
            onPress={() => setMenuVisible(true)}
          />
        }
        visible={menuVisible}
        onDismiss={() => setMenuVisible(false)}
      >
        {Object.entries(Environment).map(([key, value]) => (
          <Menu.Item
            key={key}
            title={getEnvironmentName(value)}
            onPress={() => handleEnvironmentChange(value)}
          />
        ))}
      </Menu>

      <Appbar.Content title={environmentName} />

      {showPaasServerless && (
        <Appbar.Action
          accessibilityLabel="paasserverless"
          icon="webhook"
          onPress={() => {
            const paasserverless = diagnostics?.extensions["paasserverless"];
            if (isExtensionInfo(paasserverless)) {
              onExtensionSelect(paasserverless);
            }
          }}
        />
      )}

      <Appbar.Action
        accessibilityLabel="websites"
        icon="web"
        onPress={() => {
          const websites = diagnostics?.extensions["websites"];
          if (isExtensionInfo(websites)) {
            onExtensionSelect(websites);
          }
        }}
      />
    </Appbar.Header>
  );
};

export default Header;
