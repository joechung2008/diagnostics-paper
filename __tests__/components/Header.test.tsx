import { render } from "@testing-library/react-native";
import React from "react";
import { MD3LightTheme, PaperProvider } from "react-native-paper";
import Header from "../../components/Header";
import { Environment } from "../../lib/environment";

describe("Header", () => {
  const mockOnEnvironmentChange = jest.fn();
  const mockOnExtensionSelect = jest.fn();

  it("renders correctly with basic props", () => {
    const { toJSON } = render(
      <PaperProvider theme={MD3LightTheme}>
        <Header
          diagnostics={undefined}
          environment={Environment.Public}
          onEnvironmentChange={mockOnEnvironmentChange}
          onExtensionSelect={mockOnExtensionSelect}
        />
      </PaperProvider>
    );

    expect(toJSON()).toMatchSnapshot();
  });

  it("renders correctly with diagnostics data", () => {
    const mockDiagnostics = {
      buildInfo: { buildVersion: "1.0.0" },
      extensions: {
        paasserverless: {
          extensionName: "paasserverless",
          manageSdpEnabled: true,
        },
        websites: {
          extensionName: "websites",
          manageSdpEnabled: true,
        },
      },
      serverInfo: {
        deploymentId: "test-deployment",
        extensionSync: { totalSyncAllCount: 5 },
        hostname: "test-host",
        serverId: "test-server",
      },
    };

    const { toJSON } = render(
      <PaperProvider theme={MD3LightTheme}>
        <Header
          diagnostics={mockDiagnostics}
          environment={Environment.Fairfax}
          onEnvironmentChange={mockOnEnvironmentChange}
          onExtensionSelect={mockOnExtensionSelect}
        />
      </PaperProvider>
    );

    expect(toJSON()).toMatchSnapshot();
  });
});
