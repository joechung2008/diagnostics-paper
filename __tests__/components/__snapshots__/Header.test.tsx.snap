// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly with basic props 1`] = `"Appbar.Header: [object Object],[object Object],false,[object Object]"`;

exports[`Header renders correctly with diagnostics data 1`] = `"Appbar.Header: [object Object],[object Object],[object Object],[object Object]"`;
